File: app-laravel/src/User/Application/Queries/GetUser/GetUserHandler.php
<?php

declare(strict_types=1);

namespace Src\User\Application\Queries\GetUser;

use Src\Shared\Application\Bus\Query;
use Src\Shared\Application\Bus\QueryHandler;
use Src\Shared\Domain\Exceptions\DomainException;
use Src\User\Application\DTOs\UserDTO;
use Src\User\Domain\Repositories\UserRepositoryInterface;
use Src\User\Domain\ValueObjects\UserId;

/**
 * GET USER HANDLER (CQRS Pattern)
 * ================================
 * Bu Handler GetUserQuery-ni qəbul edib istifadəçi məlumatlarını qaytarır.
 *
 * QUERY HANDLER-İN ROLU:
 * - Datanı YALNIZ oxuyur, heç vaxt dəyişmir.
 * - Domain Entity-ni Repository-dən alır.
 * - Entity-ni DTO-ya çevirir (domain obyekti xaricə çıxmamalıdır).
 * - DTO qaytarır.
 *
 * NƏYƏ ENTİTY DEYİL, DTO QAYTARIRIR?
 * - Entity-nin biznes metodları var — API-yə lazım deyil.
 * - Entity-nin daxili strukturu dəyişə bilər — API kontraktı pozulmamalıdır.
 * - Entity-də həssas data ola bilər (şifrə) — API-yə getməməlidir.
 * - DTO yalnız lazım olan dataları daşıyır (şifrə yoxdur!).
 */
final class GetUserHandler implements QueryHandler
{
    public function __construct(
        private readonly UserRepositoryInterface $userRepository,
    ) {
    }

    /**
     * İstifadəçini tap və DTO olaraq qaytar.
     *
     * @throws DomainException İstifadəçi tapılmadıqda
     */
    public function handle(Query $query): UserDTO
    {
        /** @var GetUserQuery $query */

        /**
         * String ID-ni UserId Value Object-ə çevir.
         * Əgər UUID formatı yanlışdırsa, UserId::fromString() DomainException atacaq.
         */
        $userId = UserId::fromString($query->userId);

        /**
         * Repository-dən istifadəçini axtar.
         */
        $user = $this->userRepository->findById($userId);

        if ($user === null) {
            throw new DomainException(
                "İstifadəçi tapılmadı: {$query->userId}"
            );
        }

        /**
         * Entity-ni DTO-ya çevir və qaytar.
         * UserDTO::fromEntity() yalnız lazım olan sahələri götürür (şifrəsiz!).
         */
        return UserDTO::fromEntity($user);
    }
}

File: app/spring/src/main/java/az/ecommerce/user/application/query/GetUserHandler.java
package az.ecommerce.user.application.query;

import az.ecommerce.shared.application.bus.QueryHandler;
import az.ecommerce.shared.domain.exception.EntityNotFoundException;
import az.ecommerce.user.application.dto.UserDto;
import az.ecommerce.user.domain.repository.UserRepository;
import az.ecommerce.user.domain.valueobject.UserId;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class GetUserHandler implements QueryHandler<GetUserQuery, UserDto> {

    private final UserRepository repository;

    public GetUserHandler(UserRepository repository) {
        this.repository = repository;
    }

    @Override
    @Transactional(readOnly = true, transactionManager = "userTransactionManager")
    public UserDto handle(GetUserQuery query) {
        return repository.findById(new UserId(query.userId()))
                .map(UserDto::fromDomain)
                .orElseThrow(() -> new EntityNotFoundException("User", query.userId().toString()));
    }
}
